WalletWasabi/Services/UpdateManager.cs

```diff
@@ -0,0 +1,271 @@
+using Newtonsoft.Json.Linq;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+using WalletWasabi.Microservices;
+using WalletWasabi.Models;
+using WalletWasabi.Tor.Http;
+
+namespace WalletWasabi.Services;
+
+public class UpdateManager : IDisposable
+{
+	private string InstallerName { get; set; } = "";
+	private const byte MaxTries = 3;
+	private const string ReleaseURL = "https://api.github.com/repos/zkSNACKs/WalletWasabi/releases/latest";
+
+	public UpdateManager(string dataDir, bool downloadNewVersion, IHttpClient httpClient)
+	{
+		InstallerDir = Path.Combine(dataDir, "Installer");
+		HttpClient = httpClient;
+		DownloadNewVersion = downloadNewVersion;
+	}
+
+	private async void UpdateChecker_UpdateStatusChangedAsync(object? sender, UpdateStatus updateStatus)
+	{
+		var tries = 0;
+		bool updateAvailable = !updateStatus.ClientUpToDate || !updateStatus.BackendCompatible;
+		Version targetVersion = updateStatus.ClientVersion;
+		if (!updateAvailable)
+		{
+			Cleanup();
+			return;
+		}
+		if (DownloadNewVersion)
+		{
+			do
+			{
+				tries++;
+				try
+				{
+					await GetInstallerAsync(targetVersion).ConfigureAwait(false);
+					Logger.LogInfo($"Version {targetVersion} downloaded successfuly.");
+					updateStatus.IsReadyToInstall = true;
+					break;
+				}
+				catch (Exception ex)
+				{
+					Logger.LogError($"Geting version {targetVersion} failed with error.", ex);
+				}
+			} while (tries < MaxTries);
+		}
+
+		UpdateAvailableToGet?.Invoke(this, updateStatus);
+	}
+
+	/// <summary>
+	/// Get or download installer for the newest release.
+	/// </summary>
+	/// <param name="targetVersion">This does not contains the revision number, because backend always sends zero.</param>
+	private async Task GetInstallerAsync(Version targetVersion)
```

Comment:
I think this should return installerName as a string and set it outside.

Reply:
InstallerName is set in 2 different places, it is hard to follow. Try to set class properties on the highest level can be.